builder/dockerfile/builder.go

```diff
@@ -177,13 +181,51 @@ func emitImageID(aux *streamformatter.AuxFormatter, state *dispatchState) error
 	}
 	return aux.Emit(types.BuildResult{ID: state.imageID})
 }
+func (b *Builder) dispatchDockerfileWithCancellation(parseResult *command.ParsingResult, escapeToken rune, source builder.Source) (*dispatchState, error) {
+	var state *dispatchState
+	for _, stage := range parseResult.Stages {
+		state = newDispatchState(b, escapeToken, source)
+		for i, cmd := range stage.Commands {
+			select {
+			case <-b.clientCtx.Done():
+				logrus.Debug("Builder: build cancelled!")
+				fmt.Fprint(b.Stdout, "Build cancelled")
+				buildsFailed.WithValues(metricsBuildCanceled).Inc()
+				return nil, errors.New("Build cancelled")
+			default:
+				// Not cancelled yet, keep going...
+			}
+			sourceCode := "..."
+			if src, ok := cmd.(command.WithSourceCode); ok {
+				sourceCode = src.SourceCode()
+			}
+			if len(parseResult.Stages) > 1 {
+				fmt.Fprintf(b.Stdout, "Stage %v: %v / %v %v", stage.Name, i+1, len(stage.Commands), sourceCode)
+			} else {
+				fmt.Fprintf(b.Stdout, "%v / %v %v", i+1, len(stage.Commands), sourceCode)
+			}
+			fmt.Fprintln(b.Stdout)
 
-func (b *Builder) dispatchDockerfileWithCancellation(dockerfile *parser.Result, source builder.Source) (*dispatchState, error) {
-	shlex := NewShellLex(dockerfile.EscapeToken)
-	state := newDispatchState()
-	total := len(dockerfile.AST.Children)
+			if err := b.dispatch(state, cmd); err != nil {
+				return nil, err
+			}
+
+			state.updateRunConfig()
+			fmt.Fprintf(b.Stdout, " ---> %s\n", stringid.TruncateID(state.imageID))
+			if b.options.Remove {
+				b.clearTmp()
+			}
+		}
+		if err := emitImageID(b.Aux, state); err != nil {
+			return nil, err
+		}
+	}
+	return state, nil
+}
+func (b *Builder) parseDockerfileWithCancellation(dockerfile *parser.Result, source builder.Source) (*command.ParsingResult, error) {
```

Comment:
I don't think we need cancellation for parsing. It should be a fast operation. If the context is cancelled we can pick it up before we start dispatching.